rust/src/partitions.rs

```diff
@@ -0,0 +1,100 @@
+use std::convert::TryFrom;
+
+use crate::DeltaTableError;
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum PartitionValue<T> {
+    Equal(T),
+    NotEqual(T),
+    In(Vec<T>),
+    NotIn(Vec<T>),
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct PartitionFilter<'a, T> {
+    pub partition_key: &'a str,
+    pub partition_value: PartitionValue<T>,
+}
+
+impl<'a, T: std::fmt::Debug> TryFrom<(&'a str, &str, T)> for PartitionFilter<'a, T> {
+    type Error = DeltaTableError;
+
+    fn try_from(filter: (&'a str, &str, T)) -> Result<Self, DeltaTableError> {
+        match filter {
+            (key, "=", value) if !key.is_empty() => Ok(PartitionFilter {
+                partition_key: key,
+                partition_value: PartitionValue::Equal(value),
+            }),
+            (key, "!=", value) if !key.is_empty() => Ok(PartitionFilter {
+                partition_key: key,
+                partition_value: PartitionValue::NotEqual(value),
+            }),
+            (_, _, _) => Err(DeltaTableError::InvalidPartitionFilter {
+                partition_filter: format!("{:?}", filter),
+            }),
+        }
+    }
+}
+
+impl<'a, T: std::fmt::Debug> TryFrom<(&'a str, &str, Vec<T>)> for PartitionFilter<'a, T> {
+    type Error = DeltaTableError;
+
+    fn try_from(filter: (&'a str, &str, Vec<T>)) -> Result<Self, DeltaTableError> {
+        match filter {
+            (key, "in", value) if !key.is_empty() => Ok(PartitionFilter {
+                partition_key: key,
+                partition_value: PartitionValue::In(value),
+            }),
+            (key, "not in", value) if !key.is_empty() => Ok(PartitionFilter {
+                partition_key: key,
+                partition_value: PartitionValue::NotIn(value),
+            }),
+            (_, _, _) => Err(DeltaTableError::InvalidPartitionFilter {
+                partition_filter: format!("{:?}", filter),
+            }),
+        }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct DeltaTablePartition<'a> {
+    pub key: &'a str,
+    pub value: &'a str,
+}
+
+impl<'a> DeltaTablePartition<'a> {
+    pub fn filter_accept_partition(&self, filter: &PartitionFilter<&'a str>) -> bool {
+        if filter.partition_key != self.key {
+            return true;
```

Comment:
shouldn't we return false if partition key doesn't match?

Reply:
The filters are applied with a `AND` operator on all the partition keys. So we could apply a filter with a key on a different partition key. The only case where we might be wrong is to have a filter key that doesn't exist inside the partitions keys. I will check this case before applying the filters.

> Example:
> partition_columns: "year=2020" and "month="12"
> filter: ("year"="2020") is true on the partition column "year" and also true by default with "month".
> the `all()` method is valid.

If we had:

> partition_columns: "year=2020" and "month=12"
> filter: ("year"="2020") is true on the partition column "year" and also false by default with "month".
> the `all()` method would be invalid.

Reply:
yeah, i am more concerned about error cases like typos in partition names. it looks like a more robust way to evaluate the partition predicates would be to evaluate each single filter on the full partition path, instead of each single partition component. i.e. we would need to model partition as a vector of key value pairs, then check if a filter matches any one of the key pair. And finally, we can have the evaluation logic written as:

```
for filter in all filters:
    any(filter.match(kv) for kv in partition_components )
```

Reply:
Ok, I will see how I can refactor this!

Reply:
Hum, I refactored this part based on your suggestion.

But it changes the way we apply the filters by using an `OR` operator instead of `AND`.
`(filter_1), (filter_2)` we keep partitions that matched the `filter_1` and `filter_2` by iterating over the filters.

It may not be a problem but I found easier to apply and `AND` operator with a list of filters and we may use a list of list for the OR operator. WDYT?

Reply:
> This means a partition path needs to match all the filters.
My implementation missed that part and was "a partition path needs to match one of the filters".

It's ready!